pkg/providers/openapi3/schema.go

```diff
@@ -0,0 +1,176 @@
+		}
+
+		for _, endpoints := range endpoints {
+			for name, object := range endpoints.objects() {
+				name := getCanonicalName(doc, name)
+				// empty schemaRef, skipping
+				if object == nil || object.ref == nil {
+					continue
+				}
+
+				tpl, err := newTemplate(object.ref.Value)
+
+				if err != nil {
+					return nil, fmt.Errorf("failed to build property '%s' from file '%s': %w", name, file, err)
+				}
+
+				prop := &specs.Property{
+					Template: tpl,
+					Name:     name,
+					Label:    labels.Optional,
+				}
+
+				result[prop.Name] = prop
+			}
+		}
+	}
+
+	return result, nil
+}
+
+// newTemplate builds a Property from the given swagger schemaRef component.
+// the function returns a property without name, the name should be set in the top caller.
+func newTemplate(schema *openapi.Schema) (tpl specs.Template, err error) {
+	switch schema.Type {
+	// handle as Message
+	case "object":
+		tpl, err = message(schema)
+		if err != nil {
+			return tpl, fmt.Errorf("failed to create object: %w", err)
+		}
+
+	// handle as Repeated
+	case "array":
+		tpl, err = repeated(schema)
+		if err != nil {
+			return tpl, fmt.Errorf("failed to create array: %w", err)
+		}
+
+	// check for OneOf, AnyOf, AllOf, Not
+	// and handle according to its type
+	case "":
+		switch {
+		case schema.OneOf != nil:
+			tpl, err = oneOf(schema)
+			if err != nil {
+				return tpl, fmt.Errorf("failed to create oneOf type: %w", err)
+			}
+		default:
+			return tpl, fmt.Errorf("unsupported mixed type")
+		}
+
+	// handle as scalar: string, integer, etc
+	default:
+		tpl, err = scalar(schema)
+		if err != nil {
+			return tpl, fmt.Errorf("failed to create scalar: %w", err)
+		}
+	}
+
+	return
+}
+
+// builds a message template
+func message(s *openapi.Schema) (specs.Template, error) {
+	var (
+		message = make(specs.Message, len(s.Properties))
+	)
+
+	for fieldName, ref := range s.Properties {
+		if ref.Value == nil {
+			return specs.Template{}, fmt.Errorf("field '%s' does not have schemaRef", fieldName)
+		}
+
+		tpl, err := newTemplate(ref.Value)
+		if err != nil {
+			return specs.Template{}, fmt.Errorf("failed to build field '%s': %w", fieldName, err)
+		}
+
+		message[fieldName] = &specs.Property{
+			Name:     fieldName,
+			Template: tpl,
+		}
+	}
+
+	return specs.Template{
+		Message: message,
+	}, nil
+}
+
+// builds a scalar template
+func scalar(s *openapi.Schema) (specs.Template, error) {
+	var t types.Type
+
+	switch s.Type {
+	case "string":
+		t = types.String
+	case "boolean":
+		t = types.Bool
+	case "number":
+		t = types.Float
+	case "integer":
+		t = types.Int32
+	default:
+		return specs.Template{}, fmt.Errorf("unknown type %s", s.Type)
+	}
+
+	return specs.Template{
+		Scalar: &specs.Scalar{Type: t},
+	}, nil
+}
+
+// builds a repeated template
+func repeated(s *openapi.Schema) (specs.Template, error) {
+	if s.Items == nil {
+		return specs.Template{}, errors.New("empty item schemaRef")
+	}
```

Comment:
Just wanted to check whether it is "allowed" to define a schema without items.

Reply:
it seems the package parses such schemas with no error... that's probably why I handled this case.

Reply:
This is expected behavior 🎉